server/consumer.go

```diff
@@ -1591,8 +1623,62 @@ func (o *consumer) updateConfig(cfg *ConsumerConfig) error {
 		}
 	}
 
-	// Record new config for others that do not need special handling.
-	// Allowed but considered no-op, [Description, SampleFrequency, MaxWaiting, HeadersOnly]
+	singleFilterChanged := false
+
+	multiFilterChanged := !subjectSliceEqual(cfg.FilterSubjects, o.cfg.FilterSubjects)
+
+	if cfg.FilterSubject != o.cfg.FilterSubject {
+		singleFilterChanged = true
+	}
+
+	if singleFilterChanged {
+		// We are sure that subjf len is > 0, as there is always at least one subject,
+		// even if it's a _EMPTY_
+		o.subjf[0].subject = cfg.FilterSubject
+	}
+	if multiFilterChanged {
+		// Delete old ones
+		for i, oldFilter := range o.subjf {
+			found := false
+			for _, newFilter := range cfg.FilterSubjects {
```

Comment:
Feels to me we should just wipe subjf and rebuild from the ground up like createConsumer?

Reply:
That would work, but it would mean we're refilling the buffer for each filter again.
If we retain the old `subjf` - we do not do that.

Reply:
But this is a slow path no?

Reply:
This part is, but it affects the getNextMsg (only once though) by clearing already buffered *pmsg messages for future subject sequences.

Your call if that complexity is worth one slower call to getNextMsg. Middleground would be abstracting complexity to a function and just calling it here.

Reply:
Accessing subjf is fast path, but here I thought was slow path so just rebuilding from ground up might be less error prone.

Reply:
Yes, this is a slow path, but straight-up rebuilding removes already cached messages used in hot path—the next `pmsg` per subject. Though of course, that would affect only **one** call to `getNextMsg`, so maybe I'm overoptimising for rare cases of updating filter subjects?